src/node/node.go

```diff
@@ -330,6 +341,8 @@ func (n *Node) processSyncRequest(rpc *peer.RPC, cmd *peer.SyncRequest) {
 
 	// TODO: context.Background
 	rpc.SendResult(context.Background(), n.logger, resp, respErr)
+
+	n.peerSelector.Dismiss(p.NetAddr)
```

Comment:
How about defer it after n.peerSelector.Engage(p.NetAddr) call?
(and the same for each case)